README.rst
.. contents::


Documentation
=============


What is this?
-------------

This package defines several browser views that do a redirect to
random content.


Use case
--------

The main use case would be: a random header background image.  Use
``@@randomimage`` as background image in a css file.  When the browser
loads the css file and wants to apply the background image, it will
get redirected to the image.

Example css code for when you want a random background image on all
pages except the front page, is this::

  #portal-header {
    background: url(@@randomimage) no-repeat 0 0;
  }
  .section-front-page #portal-header {
    background: #ee0;
  }


Alternative
-----------

`collective.randomheaderimage`_ achieves the same thing as using
``@@randomimage`` in a css file.  It does this by overriding the
``plone.header`` viewlet and adding some inline css to it.  This works
fine.  It has one possible drawback: if you use caching, then you
always get the same image for a page, until the cache gets refreshed.
This may or may not be a problem for you.

.. _`collective.randomheaderimage`: http://pypi.python.org/pypi/collective.randomheaderimage


Installation and configuration
------------------------------

Add ``collective.randomcontent`` to the eggs of your buildout (zcml is
not explicitly needed), rerun buildout and start your zope instance.

Install ``collective.randomcontent`` in the Site Setup.  Go to its
control panel and choose a folder in your site where we take images
from.

Note that when you only want to use ``@@randomsiteimage`` or
``@@randomsitecontent``, you do not need to install this package in
the site setup.


Details
-------

The following views are available:

- ``@@randomimage``: this picks a random image from your site and
  redirects to this url.  The image must be in a specific folder, that
  you set in the control panel.

- ``@@randomsiteimage``: this picks a random image from your site and
  redirects to this url.  The image can be anywhere in your site.

- ``@@randomcontent``: this picks a random content item from your site
  and redirects to this url.  The content must be in a specific
  folder, that you set in the control panel.

- ``@@randomsitecontent``: this picks a random content item from your
  site and redirects to this url.  The content can be anywhere in your
  site.


Compatibility
-------------

This is tested on Plone 4.2.  It will likely work on all 4.x versions.

It might work on Plone 3, but it uses ``plone.app.registry``, which is
not available by default, so you may need to take extra care about getting
versions of packages that work for your Plone version.  This is untested.


Authors
-------

- Maurits van Rees


CHANGES.txt
Changelog
=========


1.1 (unreleased)
----------------

- Nothing changed yet.


1.0 (2012-07-19)
----------------

- Initial release
  [maurits]


collective/randomcontent/tests/test_functional.py
import unittest2 as unittest

from plone.app.testing import TEST_USER_ID
from plone.app.testing import setRoles

from collective.randomcontent.testing import (
    RANDOM_CONTENT_INTEGRATION_TESTING,
    make_test_doc, make_test_image, make_test_folder, register_path,
    )


class TestRandomContent(unittest.TestCase):

    layer = RANDOM_CONTENT_INTEGRATION_TESTING

    def testRandomness(self):
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        for i in range(5):
            make_test_image(portal)
        view = portal.restrictedTraverse('randomsiteimage')
        locations = set()
        for i in range(10):
            result = view()
            self.assertFalse(result)
            response = self.layer['request'].response
            self.assertEqual(response.status, 302)
            location = response.headers.get('location')
            self.assertTrue(location)
            locations.add(location)
        # Assure that not the same image is returned each time.  Note:
        # this is expected to fail once every 10 million times.
        self.assertTrue(len(locations) > 1)

    def testCaching(self):
        # We want the content to be random, so we make sure this is
        # never cached.
        portal = self.layer['portal']
        view = portal.restrictedTraverse('randomcontent')
        result = view()
        self.assertFalse(result)
        response = self.layer['request'].response
        headers = response.headers
        self.assertEqual(headers.get('expires'),
                         'Mon, 26 Jul 1997 05:00:00 GMT')
        self.assertEqual(headers.get('pragma'), 'no-cache')
        self.assertEqual(headers.get('cache-control'),
                         'no-store, no-cache, must-revalidate')

    def testRandomSiteNoImage(self):
        portal = self.layer['portal']
        view = portal.restrictedTraverse('randomsiteimage')
        result = view()
        # There are no images, so we only get an empty string.
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 200)
        self.assertEqual(response.headers.get('location'), None)

    def testRandomSiteImage(self):
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        image = make_test_image(portal)
        view = portal.restrictedTraverse('randomsiteimage')
        result = view()
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 302)
        self.assertEqual(response.headers.get('location'),
                         image.absolute_url())

    def testRandomSiteImageNoDoc(self):
        # When searching for images, we should never find documents.
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        image = make_test_image(portal)
        make_test_doc(portal)
        view = portal.restrictedTraverse('randomsiteimage')
        locations = set()
        for i in range(10):
            result = view()
            self.assertFalse(result)
            response = self.layer['request'].response
            self.assertEqual(response.status, 302)
            location = response.headers.get('location')
            self.assertTrue(location)
            locations.add(location)
        self.assertEqual(len(locations), 1)
        self.assertEqual(locations.pop(),
                         image.absolute_url())

    def testRandomImageNoFolder(self):
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        image = make_test_image(portal)
        view = portal.restrictedTraverse('randomimage')
        result = view()
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 302)
        self.assertEqual(response.headers.get('location'),
                         image.absolute_url())

    def testRandomImageWithEmptyFolder(self):
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        make_test_image(portal)
        make_test_folder(portal, register=True)
        view = portal.restrictedTraverse('randomimage')
        result = view()
        # There are no images, so we only get an empty string.
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 200)
        self.assertEqual(response.headers.get('location'), None)

    def testRandomImage(self):
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        make_test_image(portal)
        folder = make_test_folder(portal, register=True)
        folder_image = make_test_image(folder)
        view = portal.restrictedTraverse('randomimage')
        result = view()
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 302)
        self.assertEqual(response.headers.get('location'),
                         folder_image.absolute_url())
        # Let's see if that is not a lucky break.
        locations = set()
        for i in range(10):
            result = view()
            self.assertFalse(result)
            response = self.layer['request'].response
            self.assertEqual(response.status, 302)
            location = response.headers.get('location')
            self.assertTrue(location)
            locations.add(location)
        self.assertEqual(len(locations), 1)

    def testRandomImageNoDoc(self):
        # When searching for images, we should never find documents.
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        make_test_image(portal)
        make_test_doc(portal)
        folder = make_test_folder(portal, register=True)
        folder_image = make_test_image(folder)
        make_test_doc(folder)
        view = portal.restrictedTraverse('randomimage')
        locations = set()
        for i in range(10):
            result = view()
            self.assertFalse(result)
            response = self.layer['request'].response
            self.assertEqual(response.status, 302)
            location = response.headers.get('location')
            self.assertTrue(location)
            locations.add(location)
        self.assertEqual(len(locations), 1)
        self.assertEqual(locations.pop(),
                         folder_image.absolute_url())

    def testRandomSiteContent(self):
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        doc = make_test_doc(portal)
        view = portal.restrictedTraverse('randomsitecontent')
        result = view()
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 302)
        self.assertEqual(response.headers.get('location'),
                         doc.absolute_url())

    def testRandomContent(self):
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        make_test_doc(portal)
        folder = make_test_folder(portal, register=True)
        # Note: the folder itself is an excellent random content item,
        # so we do not need to add another document.
        #folder_doc = make_test_doc(folder)
        view = portal.restrictedTraverse('randomcontent')
        result = view()
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 302)
        self.assertEqual(response.headers.get('location'),
                         folder.absolute_url())
        # Let's see if that is not a lucky break.
        locations = set()
        for i in range(10):
            result = view()
            self.assertFalse(result)
            response = self.layer['request'].response
            self.assertEqual(response.status, 302)
            location = response.headers.get('location')
            self.assertTrue(location)
            locations.add(location)
        self.assertEqual(len(locations), 1)

    def testRandomSiteImageOtherFolder(self):
        # Check that we can use the views (or one of them anyway) on a
        # regular folder.
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        image = make_test_image(portal)
        folder = make_test_folder(portal)
        view = folder.restrictedTraverse('randomsiteimage')
        result = view()
        self.assertFalse(result)
        response = self.layer['request'].response
        self.assertEqual(response.status, 302)
        self.assertEqual(response.headers.get('location'),
                         image.absolute_url())

    def testRandomSiteImageNavigationRoot(self):
        # Searching for an image in the entire site still takes the
        # navigation root into account.
        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        make_test_image(portal)
        folder = make_test_folder(portal)
        from plone.app.layout.navigation.interfaces import INavigationRoot
        from zope.interface import alsoProvides
        alsoProvides(folder, INavigationRoot)
        folder_image = make_test_image(folder)
        view = folder.restrictedTraverse('randomsiteimage')
        locations = set()
        for i in range(10):
            result = view()
            self.assertFalse(result)
            response = self.layer['request'].response
            self.assertEqual(response.status, 302)
            location = response.headers.get('location')
            self.assertTrue(location)
            locations.add(location)
        self.assertEqual(len(locations), 1)
        self.assertEqual(locations.pop(),
                         folder_image.absolute_url())

    def test_path(self):
        # Test some corner cases.
        def get_path(obj):
            return '/'.join(obj.getPhysicalPath())

        portal = self.layer['portal']
        setRoles(portal, TEST_USER_ID, ('Manager',))
        nav_root = make_test_folder(portal)
        from plone.app.layout.navigation.interfaces import INavigationRoot
        from zope.interface import alsoProvides
        alsoProvides(nav_root, INavigationRoot)

        portal_view = portal.restrictedTraverse('randomcontent')
        nav_root_view = nav_root.restrictedTraverse('randomcontent')
        self.assertEqual(portal_view._get_nav_root_path(),
                         get_path(portal))
        self.assertEqual(nav_root_view._get_nav_root_path(),
                         get_path(nav_root))
        self.assertEqual(portal_view._get_folder_path(),
                         get_path(portal))
        self.assertEqual(nav_root_view._get_folder_path(),
                         get_path(nav_root))

        top_folder = make_test_folder(portal, new_id='top', register=True)
        self.assertEqual(portal_view._get_folder_path(),
                         get_path(top_folder))
        self.assertEqual(nav_root_view._get_folder_path(),
                         get_path(top_folder))
        # The path does not need to be the full path from the zope
        # root.
        register_path('top')
        self.assertEqual(portal_view._get_folder_path(),
                         get_path(top_folder))
        self.assertEqual(nav_root_view._get_folder_path(),
                         get_path(top_folder))

        # Make two folders with the same id
        folder_in_portal = make_test_folder(portal, 'target')
        folder_in_nav_root = make_test_folder(nav_root, 'target')
        register_path('target')
        self.assertEqual(portal_view._get_folder_path(),
                         get_path(folder_in_portal))
        self.assertEqual(nav_root_view._get_folder_path(),
                         get_path(folder_in_nav_root))

        # Try with the top level folder again, calling the view on a
        # folder at a deeper level.
        register_path('top')
        view = folder_in_nav_root.restrictedTraverse('randomcontent')
        self.assertEqual(view._get_folder_path(),
                         get_path(top_folder))

        # If the path cannot be traversed, so be it.
        register_path('non-existing')
        self.assertEqual(portal_view._get_folder_path(),
                         get_path(portal))
        self.assertEqual(nav_root_view._get_folder_path(),
                         get_path(nav_root))


collective/randomcontent/testing.py
from plone.app.testing import IntegrationTesting
from plone.app.testing import PLONE_FIXTURE
from plone.app.testing import PloneSandboxLayer
from plone.app.testing import applyProfile
from plone.registry.interfaces import IRegistry
from zope.component import getUtility
from zope.configuration import xmlconfig


class RandomContentLayer(PloneSandboxLayer):

    defaultBases = (PLONE_FIXTURE,)

    def setUpZope(self, app, configurationContext):
        # Load ZCML
        import collective.randomcontent
        xmlconfig.file('configure.zcml', collective.randomcontent,
                      context=configurationContext)

    def setUpPloneSite(self, portal):
        # Apply our profile.
        applyProfile(portal, 'collective.randomcontent:default')

RANDOM_CONTENT_FIXTURE = RandomContentLayer()
RANDOM_CONTENT_INTEGRATION_TESTING = IntegrationTesting(
    bases=(RANDOM_CONTENT_FIXTURE,), name="RandomContent:Integration")

# tool.gif
IMAGE_DATA = (
    'GIF89a\x10\x00\x10\x00\xf2\x08\x00>\xa8\xd2\x12\x9a\xcb*\xa2\xcf\xac\xd7'
    '\xebP\xaf\xd5a\xb6\xd9\x81\xc3\xe1\x00\x93\xc8!\xff\x0bNETSCAPE2.0\x03'
    '\x01\x00\x00\x00!\xf9\x04\x05\x00\x00\x08\x00,\x00\x00\x00\x00\x10\x00'
    '\x10\x00\x00\x04Z\x10I4\xc8\xb9\x87\x8c\xc9\x01&\xd6\x05pHp\x08\xdb4'
    '\x08G0y\x85d\x18\x9c5\x0e\xa7|\xd1\x13[\x1d)\xc3\x8e\x83\x0bqf2\xde\x87'
    '\xa4;\xd0B\x04&B\xe8D\x84\x0eG\x1e\x12q\xb1\x04\x87\xaa\x0cN\xd0\xe4I|'
    '\x08Otj\xb6\x1eF\x12\x13\x8a\xc8ru>W8\xe9\xf7II"\x00;')

# A few helper functions.


def make_test_image(context):
    new_id = context.generateUniqueId('Image')
    context.invokeFactory('Image', new_id, image=IMAGE_DATA)
    image = context[new_id]
    image.reindexObject()  # Might have already happened, but let's be sure.
    return image


def make_test_doc(context):
    new_id = context.generateUniqueId('Document')
    context.invokeFactory('Document', new_id)
    doc = context[new_id]
    doc.reindexObject()  # Might have already happened, but let's be sure.
    return doc


def register_path(path):
    # Register as random content folder
    registry = getUtility(IRegistry)
    randomContentFolder = registry.records.get(
        'collective.randomcontent.interfaces.IRandomContentSettings'
        '.randomContentFolder')
    randomContentFolder.value = path


def make_test_folder(context, new_id=None, register=False):
    if not new_id:
        new_id = context.generateUniqueId('Folder')
    context.invokeFactory('Folder', new_id)
    folder = context[new_id]
    folder.reindexObject()
    if register:
        # Register as random content folder
        register_path('/'.join(folder.getPhysicalPath()))
    return folder


collective/randomcontent/vocabularies.py
from Products.CMFCore.utils import getToolByName
from zope.component.hooks import getSite
from zope.schema.interfaces import IVocabularyFactory
from zope.schema.vocabulary import SimpleVocabulary, SimpleTerm
from zope.interface import implements


class PortalTypesVocabulary(object):

    implements(IVocabularyFactory)

    def __init__(self, portal_type):
        self.portal_type = portal_type

    def __call__(self, context):
        site = getSite()
        catalog = getToolByName(site, 'portal_catalog', None)
        if catalog is None:
            return SimpleVocabulary([])

        brains = catalog(portal_type=self.portal_type)
        items = [(brain.UID, brain.Title) for brain in brains]
        # Most schema fields expect unicode values.
        terms = [SimpleTerm(value=unicode(pair[0]), token=unicode(pair[0]),
                             title=pair[1]) for pair in items]
        return SimpleVocabulary(terms)


# Products.CMFCore.interfaces._content.IFolderish
FolderVocabularyFactory = PortalTypesVocabulary('Folder')


collective/randomcontent/browser.py
import random

from Acquisition import aq_inner
from Products.CMFCore.utils import getToolByName
from Products.Five import BrowserView
from plone.registry.interfaces import IRegistry
from zope.component import getUtility
from zope.component import getMultiAdapter


class BaseRandomView(BrowserView):

    portal_type = None

    def _nocache(self):
        """Don't you dare cache this ever.

        It should be a redirect so it probably does not get cached
        anyway, but let's be sure.  Also, if no images are found, no
        redirect takes place.
        """
        self.request.response.setHeader(
            'Expires', 'Mon, 26 Jul 1997 05:00:00 GMT')
        self.request.response.setHeader(
            'Cache-Control', 'no-store, no-cache, must-revalidate')
        self.request.response.setHeader(
            'Pragma', 'no-cache')

    def _get_nav_root_path(self):
        """Get the navigation root path.
        """
        context = aq_inner(self.context)
        pps = getMultiAdapter((context, self.request),
                              name='plone_portal_state')
        return pps.navigation_root_path()

    def _get_folder_path(self):
        """Get the path of the folder selected for random content.
        """
        context = aq_inner(self.context)
        pps = getMultiAdapter((context, self.request),
                              name='plone_portal_state')
        root_path = pps.navigation_root_path()
        registry = getUtility(IRegistry)
        randomContentFolder = registry.records.get(
            'collective.randomcontent.interfaces.IRandomContentSettings'
            '.randomContentFolder')
        if not randomContentFolder or not randomContentFolder.value:
            return root_path
        content_path = randomContentFolder.value
        portal = pps.portal()
        if content_path.startswith(root_path):
            path = content_path
        else:
            path = '/'.join([root_path, content_path])
        path = path.replace('//', '/')
        target = portal.unrestrictedTraverse(path, None)
        if target:
            # Acquisition could mean the target has a different
            # physical path.
            return '/'.join(target.getPhysicalPath())
        return root_path

    def _get_path(self):
        """Get the path in which we will look for random content.

        By default we use the navigation root path.
        """
        return self._get_nav_root_path()

    def _update_filter(self, myfilter):
        """Update the filter.

        By default we do nothing.  Override this if you need special
        handling.  The filter is passed to the portal_catalog.
        """
        pass

    def _filter(self):
        myfilter = {}
        path = self._get_path()
        if path:
            myfilter['path'] = path
        if self.portal_type:
            myfilter['portal_type'] = self.portal_type
        self._update_filter(myfilter)
        return myfilter

    def __call__(self):
        self._nocache()
        context = aq_inner(self.context)
        catalog = getToolByName(context, 'portal_catalog')
        brains = catalog(**self._filter())
        if not brains:
            return ""
        rnd = random.randint(0, len(brains) - 1)
        self.request.RESPONSE.redirect(brains[rnd].getURL())
        return ""


class RandomSiteContent(BaseRandomView):
    pass


class RandomContent(BaseRandomView):

    def _get_path(self):
        """Get the path in which we will look for random content.

        We want to look in the selected folder.
        """
        return self._get_folder_path()


class RandomSiteImage(BaseRandomView):

    portal_type = 'Image'


class RandomImage(BaseRandomView):

    portal_type = 'Image'

    def _get_path(self):
        """Get the path in which we will look for random content.

        We want to look in the selected folder.
        """
        return self._get_folder_path()
